README.md
# pimatic-location

To use this app you need to install the Plugin "pimatic-location".
You need to add a device to your Pimatic config for each smartphone that should report it's location.

Plugin:

Simply add 
```
    {
      "plugin": "location"
    },
```
to your plugin configuration.

For each Device (Smartphone) you want to 'track', you need to create a corresponding device in your config.
Example:
```
    {
      "id": "your-phone",
      "name": "your-phone",
      "class": "LocationDevice",
      "lat": 52.5200066,
      "long": 13.404954
    },
```

If you want to use the Apple iCloud to locate your Devices your Device-Configuration should look like this:
```
    {
      "id": "your-phone",
      "name": "your-phone",
      "class": "LocationDevice",
      "lat": 52.5200066,
      "long": 13.404954,
      "iCloudUser": "Username",
      "iCloudPass": "Password",
      "iCloudDevice": "DeviceName",
      "iCloudInterval": 60000
    },
```

Optional parameters for each Device:
```
      "useGoogleMaps": true,
      "googleMapsApiKey": "your-api-key-here"
```
With the optional parameter "useGoogleMaps" you can optionaly disable the usage of GoogleMaps. (You no longer get a route-distance, eta or address)
With the parameter "googleMapsApiKey" you can define your own ApiKey, if you got to many googleMapsApi-requests running.
You can obtain a ApiKey in the Google developers console https://console.developers.google.com . You need also to activate the "Directions API" in your account.

The lat and long values correspond to the location you want the distance to be calculated.
You can use this website to get your longitude and latitude.
http://www.mapcoordinates.net/en

# Configuring the Application
Following Options are available:

- Host: The IP or Hostname of your Pimatic.
- Protocol: The Protocol to use to connect to Pimatic. (Default: http)
- Port: The Port to use to connect to Pimatic. (Default: 80)
- User: The Pimatic-User to use to connect to your Pimatic. (Default: Admin)
- Password: The Password corresponding to the Pimatic-User.
- Device-ID: The id of the Device you created for this Smartphone.
- Interval: The interval of location updates in milliseconds at which you prefer to receive location updates. (Default: 600000 (10 Minutes))
- Interval Limit: The lower limit of location updates in milliseconds you want to receive. (Default: 60000 (1 Minute))
- Priority: The priority of the location updates you want to receive. (Default: Balanced Power)
  * Balanced Power: Precision to within a city block, which is an accuracy of approximately 100 meters. With this setting, the location services are likely to use   * WiFi and cell tower positioning.
  * High: Request the most precise location possible. With this setting, the location services are more likely to use GPS.
  * Low: Use this setting to request city-level precision, which is an accuracy of approximately 10 kilometers.
  * No: Use this setting if you need negligible impact on power consumption, but want to receive location updates when available. With this setting, the app does not trigger any location updates, but receives locations triggered by other apps.

(Source: https://developer.android.com/training/location/receive-location-updates.html#location-request)
- Auto refresh: Starts the service that updates the smartphones location in background, based on your settings, to your Pimatic. (Default: On)
- Write logfile: Write to a logfile informations about location updates and service activity. (Default: Off)
- Report current address: Should your current address be shown in pimatic?
- Test & Save: Test and Save your Settings. Starts the location-Service if auto refresh is on.

# Tasker integration

You can use the Android App "Tasker" to send an intent to the application to set a new interval or a new priority.

Tasker example intent config for interval:
```
Send Intent [ 
 Action:android.intent.action.SEND 
 Cat:Default 
 Mime Type:text/plain 
 Data: 
 Extra:android.intent.extra.INTERVAL: 600000
 Extra: 
 Package:de.blackoise.pimaticlocation
 Class: 
 Target:Service ]
```

Tasker example intent config for interval limit:
```
Send Intent [ 
 Action:android.intent.action.SEND 
 Cat:Default 
 Mime Type:text/plain 
 Data: 
 Extra:android.intent.extra.INTERVALLIMIT: 600000
 Extra: 
 Package:de.blackoise.pimaticlocation
 Class: 
 Target:Service ]
```

Tasker example intent config for priority:
```
Send Intent [ 
 Action:android.intent.action.SEND 
 Cat:Default 
 Mime Type:text/plain 
 Data: 
 Extra:android.intent.extra.PRIORITY: 0
 Extra: 
 Package:de.blackoise.pimaticlocation
 Class: 
 Target:Service ]
```

Priorities:
0 = Balanced Power
1 = High
2 = Low
3 = No

# Screenshots
![Settings Window](https://raw.githubusercontent.com/Oitzu/pimatic-location/master/Android/screenshots/2015-02-06%2012.05.59.png)

changelog
Version 1.1.3
===============
* Start service on device boot

Version 1.1.2
===============
* Replaced switch button with checkbox for backward compatibility.
* Raised targetSdkVersion to 22 (Android 5.1.1).

Version 1.1.1
===============
* Added French translation (thx to sebastien-vg)
* Fixed bug "App crashes if location services disabled" #18
* Fixed bug "Prevent a crash by NPE in service if intent == null" #24 (thx to sweetpi)

Version 1.1.0
===============
* Changed location updates to Google Services API.
* Integrated use of pimatic-location plugin.
* Changed to send location instead distance to plugin. Calculation moved to plugin.
* Added the option to decide if the current address should be shown in Pimatic. (Privacy option)
* Added the option to change the interval limit.
* Fixed crash on no connection bug.
* Added translation feature.

Version 1.0.1
===============
* Added logfile and logfile view
* Added intent to change interval
* Changed to location listener

Android/PimaticLocation/app/src/main/java/de/blackoise/pimaticlocation/PLService.java
package de.blackoise.pimaticlocation;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.location.Location;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;
import com.loopj.android.http.JsonHttpResponseHandler;
import com.google.android.gms.location.LocationListener;

import org.apache.http.Header;
import org.apache.http.conn.ConnectionReleaseTrigger;
import org.json.JSONException;
import org.json.JSONObject;
import java.io.FileOutputStream;

/**
 * Created by Oitzu on 03.02.2015.
 */
public class PLService extends Service implements GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener, LocationListener {
    private LocationManager locManager;
    private LocationRequest mLocationRequest;
    private GoogleApiClient mGoogleApiClient;

    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onConnectionSuspended(int cause) {
        Log.i("PLService", "Connection to GoogleApiClient suspended");
        mGoogleApiClient.connect();
    }

    @Override
    public void onCreate() {

        Log.v("PLService", "Service created.");
    }

    @Override
    public void onConnected(Bundle connectionHint)
    {
        Log.v("PLService", "Connected to GoogleApiClient.");

        LocationServices.FusedLocationApi.requestLocationUpdates(
                mGoogleApiClient, mLocationRequest, this);
    }

    @Override
    public void onConnectionFailed(ConnectionResult result)
    {
        Log.i("PLService", "Connection to GoogleApiClient failed: ConnectionResult.getErrorCode() = " + result.getErrorCode());
    }

    @Override
    public void onLocationChanged(Location location) {

        if(location!=null){
            writeLog("Location update received. Provider: " + location.getProvider());
            updateLocation(location);
            Log.v("Debug", "Location changed.");
        }
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.v("PLService", "Service started.");
        final SharedPreferences settings = getSharedPreferences("de.blackoise.pimaticlocation", MODE_PRIVATE);
        if (intent != null && Intent.ACTION_SEND.equals(intent.getAction()) && intent.getType() != null) {
            writeLog("Intent with type " + intent.getType()+" received.");
            if ("text/plain".equals(intent.getType())) {
                Bundle extras = intent.getExtras();
                if(extras.containsKey("android.intent.extra.INTERVAL"))
                {
                    settings.edit().putString("Interval",String.valueOf(extras.getInt("android.intent.extra.INTERVAL"))).apply();
                    writeLog("Setting new Interval "+ String.valueOf(extras.getInt("android.intent.extra.INTERVAL")) + " by intent.");
                }
                if(extras.containsKey("android.intent.extra.PRIORITY"))
                {
                    settings.edit().putInt("Priority", extras.getInt("android.intent.extra.PRIORITY")).apply();
                    writeLog("Setting new Priority "+ String.valueOf(extras.getInt("android.intent.extra.PRIORITY")) + " by intent.");
                }
                if(extras.containsKey("android.intent.extra.INTERVALLIMIT"))
                {
                    settings.edit().putInt("IntervalLimit", extras.getInt("android.intent.extra.INTERVALLIMIT")).apply();
                    writeLog("Setting new Interval Limit "+ String.valueOf(extras.getInt("android.intent.extra.INTERVALLIMIT")) + " by intent.");
                }
            }
        }

        if(mGoogleApiClient != null)
        {
            if(mGoogleApiClient.isConnected()) {
                mGoogleApiClient.disconnect();
            }
        }

        //mGoogleApiClient.connect();

        mGoogleApiClient = new GoogleApiClient.Builder(this)
                .addConnectionCallbacks(this)
                .addOnConnectionFailedListener(this)
                .addApi(LocationServices.API)
                .build();


        int Interval = Integer.parseInt(settings.getString("Interval", "600000"));
        int IntervalLimit = Integer.parseInt(settings.getString("IntervalLimit", "60000"));

        mLocationRequest = new LocationRequest();
        mLocationRequest.setInterval(Interval);
        mLocationRequest.setFastestInterval(IntervalLimit);

        String accuracy;

        switch(settings.getInt("Priority", 0))
        {
            case 0:
                mLocationRequest.setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY);
                accuracy = "Balanced Power";
                break;
            case 1:
                mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
                accuracy = "High";
                break;
            case 2:
                mLocationRequest.setPriority(LocationRequest.PRIORITY_LOW_POWER);
                accuracy = "Low";
                break;
            case 3:
                mLocationRequest.setPriority(LocationRequest.PRIORITY_NO_POWER);
                accuracy = "No";
                break;
            default:
                mLocationRequest.setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY);
                accuracy = "Balanced Power";
        }

        if(mGoogleApiClient.isConnected())
        {
            mGoogleApiClient.reconnect();
        }
        else
        {
            mGoogleApiClient.connect();
        }

        writeLog("Starting service with interval of " + Integer.parseInt(settings.getString("Interval", "600000")) +"ms and " + accuracy + " Accuracy.");

        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        if (mGoogleApiClient.isConnected()) {
            mGoogleApiClient.disconnect();
        }
        super.onDestroy();
    }

    private void updateLocation(final Location lastKnownLocation)
    {
        final SharedPreferences settings = getSharedPreferences("de.blackoise.pimaticlocation", MODE_PRIVATE);
        final API api = new API(settings.getString("Host", "pimatic.example.org"), settings.getString("Protocol", "http"), settings.getString("Port", "80"), settings.getString("User", "admin"), settings.getString("Password", "admin"));

        try {
            //update location
            JSONObject jsonParams = new JSONObject();
            jsonParams.put("long", lastKnownLocation.getLongitude());
            jsonParams.put("lat", lastKnownLocation.getLatitude());
            jsonParams.put("updateAddress", settings.getBoolean("reportAddress",false)?1:0);

            api.update_Location(settings.getString("DeviceID", android.os.Build.MODEL), getApplicationContext(), jsonParams, new JsonHttpResponseHandler(){
                @Override
                public void onSuccess(int statusCode, Header[] headers, JSONObject response) {
                    writeLog("Updated location successfully.");
                }

                @Override
                public void onFailure(int statusCode, Header[] headers, Throwable e, JSONObject response) {
                    writeLog("Failed to update location. Check your connection.");
                }
            });
        }
        catch (JSONException e)
        {
            e.printStackTrace();
        }
    }



    private void writeLog(String text)
    {
        final SharedPreferences settings = getSharedPreferences("de.blackoise.pimaticlocation", MODE_PRIVATE);
        if(settings.getBoolean("writeLogfile", true))
        {
            try {
                String date = android.text.format.DateFormat.format("yyyy-MM-dd hh:mm:ss", new java.util.Date()).toString();

                String logLine = date + ": " + text + "\n";

                FileOutputStream fos = openFileOutput("logfile", Context.MODE_APPEND);

                fos.write(logLine.getBytes());

                fos.close();

            } catch (Exception e) {

                e.printStackTrace();

            }
        }
    }
}
